src/python/dart/engine/redshift/actions/consume_subscription.py

```diff
@@ -26,10 +28,27 @@ def consume_subscription(redshift_engine, datastore, action):
         _logger.info('setting up schemas, tables, s3_path generator')
         create_schemas_and_tables(conn, action, dataset)
         create_tracking_schema_and_table(conn, action)
-        s3_path = get_most_recently_processed_s3_path(conn, action)
-
-        s3_path_and_updated_generator = _s3_path_and_updated_generator(dart, subscription.id, action.id, s3_path)
-        copy_from_s3(dart, datastore, action, dataset, conn, batch_size, s3_path_and_updated_generator)
+        nudge_batches = None
+        if nudge_id:
+            most_recent_batch_id = get_most_recently_processed_nudge_batch(conn, action)
+            nudge_batches = dart.get_latest_nudge_batches(nudge_id, most_recent_batch_id)
+            if not nudge_batches:
+                new_batch = dart.create_nudge_batch(nudge_id)
+                if new_batch['BatchId']:
+                    nudge_batches = [new_batch['BatchId']]
+            s3_path_and_updated_generator = _nudge_s3_path_and_updated_generator(dart,
+                                                                                 subscription.nudge_id,
+                                                                                 nudge_batches)
+        else:
+            most_recent_s3_path = get_most_recently_processed_s3_path(conn, action)
+            s3_path_and_updated_generator = _s3_path_and_updated_generator(dart,
+                                                                           subscription.id,
+                                                                           action.id,
+                                                                           most_recent_s3_path)
+        copy_from_s3(dart, datastore, action, dataset, conn, batch_size, s3_path_and_updated_generator, nudge_id)
+        if nudge_batches:
+            for b in nudge_batches:
```

Comment:
You're going to be re-consuming batches if they're replayed. I return a 'State' field. Consume if state == 'UNCONSUMED'?